disco/disco.go

```diff
@@ -45,22 +55,62 @@ type Disco struct {
 	hostCache map[svchost.Hostname]*Host
 	credsSrc  auth.CredentialsSource
 
+	// discoTimeout is the timeout in seconds for the discovery request.
+	// The default of 11 seconds is used unless a client overrides the value.
+	discoTimeout time.Duration
+
 	// Transport is a custom http.RoundTripper to use.
 	Transport http.RoundTripper
 }
 
+// ErrServiceDiscoveryNetworkRequest represents the error that occurs when
+// the service discovery fails for an unknown network problem.
+type ErrServiceDiscoveryNetworkRequest struct {
+	err error
+}
+
+func (e *ErrServiceDiscoveryNetworkRequest) Error() string {
+	return e.err.Error()
+}
+
 // New returns a new initialized discovery object.
 func New() *Disco {
 	return NewWithCredentialsSource(nil)
 }
 
+func retryableTransport() http.RoundTripper {
+	if client == nil {
+		client = &http.Client{
+			Timeout: defaultDiscoveryTimeout,
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				log.Printf("[DEBUG] Service discovery redirected to %s", req.URL)
+				if len(via) > maxRedirects {
+					return errors.New("too many redirects") // this error will never actually be seen
+				}
+				return nil
+			},
+		}
+	}
+	retryableClient := retryablehttp.NewClient()
+	retryableClient.HTTPClient = client
+	retryableClient.RetryMax = defaultDiscoveryRetry
+	retryableClient.CheckRetry = retryHTTPCheck
+
+	return retryableClient.HTTPClient.Transport
+}
+
+func retryHTTPCheck(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	return true, nil
+}
```

Comment:
What are some of the implications of always setting retry = yes? Digging into the DefaultRetryPolicy from retryablehttp, it looks like it's skipping retries for things like circular redirects, invalid protocols like `hppt:`, invalid certs, etc. — that all seems desirable to me, but I admit I'm still getting a grip on the problem space here.

Reply:
After running some tests this afternoon, I remember why I made this change (& I wish past me had added a comment). With a slow network, one could see context.DeadlineExceeded and context.Canceled errors, so I tried setting the method to retry with any errors to see what happened.

I'm going to add another retry method in the `go-retryablehttp` that has the logic that I think we're looking for.